packages/vite/scripts/emitCjsTypes.ts

```diff
@@ -0,0 +1,96 @@
+import fs from 'node:fs'
+import path from 'node:path'
+import glob from 'fast-glob'
+import * as lexer from 'es-module-lexer'
+import colors from 'picocolors'
+
+async function main() {
+  const typeFiles = await glob('**/*.d.ts', {
+    cwd: 'dist',
+    absolute: true
+  })
+
+  for (const file of typeFiles) {
+    let text = fs.readFileSync(file, 'utf8')
+
+    const [imports] = lexer.parse(text)
+    const typeImports = parseTypeImports(text, true)
+    const allImports = [...imports, ...typeImports]
+      // In reverse order
+      .sort((a, b) => b.s - a.s)
+
+    for (const i of allImports) {
+      let id = i.n
+      if (!id || !/^\.\.?(?:\/|$)/.test(id) || id.endsWith('.cjs')) {
+        continue
+      }
+
+      const importedFile = path.resolve(path.dirname(file), id)
+      if (isDirectory(importedFile)) {
+        id += '/index'
+      }
+
+      const cjsModuleSpec = id + '.cjs'
+      text = text.slice(0, i.s) + cjsModuleSpec + text.slice(i.e)
```

Comment:
So this adds `.cjs` to all file imports, do we need to do the same for `node/index.d.ts` to with `.js` only. I've seen this need in the past with node16, but i've also heard that it's not exactly needed.

I also wonder if adding `.js` is enough to get CJS support? so we avoid another file.

Reply:
I've tested explicit `.js` in `node/index.d.ts` before taking this route. It doesn't work, unfortunately.